### PR TITLE
WMS service: option to skip name attribute for groups in wms capabilities document

### PR DESCRIPTION
**Pull request by tomass**
## Description

When QGIS project contains layer groups, they are listed in WMS capabilities document alongside with layers. If a group (it's name as listed in capabilities) is included in WMS LAYERS alongside with names of layers in that group, QGIS would duplicate the layers: once for the group in LAYERS= and once for specific layer in LAYERS= Here is an example of result with duplicate layers - note the duplicate labels:
image

OGC WMS specification states that name attribute can be skipped. Technique of skipping name tag but leaving all other information including the title tag is used in say geoserver to avoid adding a group in LAYERS= but still being able to display the group in layer hierarchy.

## Change

I've added a project level setting "Skip name attribute for groups" which by default is not checked leaving the existing behaviour - groups do have a name attribute. This means functionality will not change for existing deployments.

If this setting is checked it will not add a name attribute for groups in WMS capabilities document.

Example excerpt from GetCapabilities response:

```
<Layer queryable="1">
  <Name>layer_name</Name> <--- this will be skipped when option is enabled
  <Title>Human readable name</Title>
```
## Our practice

We had a problem with duplicate layers when groups were included in LAYERS= of WMS request. Therefore we've been doing such change (removing of name attribute) manually from capabilities xml for several months with WMS clients using our services working as expected in QGIS as well as other on-line and desktop GIS clients.

### CODE REVIEW

**Review by Gustry — APPROVED** [review]
Review (about UI and tests) done in #54146

**Comment by Gustry** [comment]
Someone having commit rights on the QGIS server side : @pblottiere @rldhont @elpaso or @mhugent

**Review by rldhont — APPROVED** [review]

**Comment by rldhont** [comment]
Waiting for an other approval from @pblottiere @elpaso or @mhugent for merging

**Review by elpaso — COMMENTED** [review]
Can you please:

- replace QGZ with plain QGS files?
- confirm that when the group name is skipped a client request for a layer (group) with that name would NOT be honored?

**Comment by tomass** [comment]
@elpaso OK, I can replace QGZ with QGS.
this option has no influence on actual GetMap requests. Only GetCapabilities response is altered. Idea is that when client requests the names of layers via GetCapabilities, he does not get a name for a group - therefore he is unable to know what name to request for a group in GetMap. This works for web clients we've checked this in production. And has minimal impact on existing code.

**Comment by elpaso** [comment]
> @elpaso OK, I can replace QGZ with QGS. this option has no influence on actual GetMap requests. Only GetCapabilities response is altered. Idea is that when client requests the names of layers via GetCapabilities, he does not get a name for a group - therefore he is unable to know what name to request for a group in GetMap. This works for web clients we've checked this in production. And has minimal impact on existing code.

Yeah, I suspected that you left that out.

My reasoning is that whenever you send a request with LAYERS that do not exist or are not allowed you expect to get back an error but with this PR this isn't true anymore.

So I am asking you to add the logic (and a test) to exclude group names from the allowed LAYERS when the flag is set.

`?REQUEST=GetMap&LAYERS=<my excluded group name> ` must return an error.

**Comment by elpaso** [comment]
@tomass you'll probably need to hack into `void QgsWmsRenderContext::initLayerGroupsRecursive`

**Comment by tomass** [comment]
@elpaso  done.
Currently error message about undefined layers in LAYERS= parameter is thrown from _searchLayersToRenderSld_ and _searchLayersToRenderSld_, so I've added additional check in the same functions.
I would have liked _isValidLayer_ and/or _isValidGroup_ better, but those are never called (at least in my tests).

File: tests/src/python/test_qgsserver_wms_getcapabilities_group_name.py
Comment (elpaso) [inline tests/src/python/test_qgsserver_wms_getcapabilities_group_name.py:37]: For testing inclusion or exclusion of a string (or bytes) in the response you can use this construct directly:

```suggestion
        self.assertIn(b'<Name>layer_group</Name>', r, 'attribute <name> should be specified for a layer group')
```

no need to use find.  Conversion to str from bytes can also be avoided.

Same for the other tests.

File: tests/src/python/CMakeLists.txt
Comment (elpaso) [inline tests/src/python/CMakeLists.txt:590]:
```suggestion
  ADD_PYTHON_TEST(PyQgsServerWMSGetCapabilitiesGroupName test_qgsserver_wms_getcapabilities_group_name.py)
```

File: tests/src/python/test_qgsserver_wms_getcapabilities_group_name.py
Comment (elpaso) [inline tests/src/python/test_qgsserver_wms_getcapabilities_group_name.py:3]:
```suggestion
From build dir, run: ctest -R PyQgsServerWMSGetCapabilitiesGroupName -V
```

**Comment by elpaso** [comment]
@tomass looks good to me now, thank you.

I would probably have avoided to create a dedicated/separated test for getcap, most of getcap tests for WMS are in `src/python/test_qgsserver_wms.py` and you could have added your tests to that file but this is not a problem if you want to keep them separated, maybe just rename it as suggested, your choice.

File: tests/src/python/test_qgsserver_wms_getcapabilities_group_name.py
Comment (tomass) [reply tests/src/python/test_qgsserver_wms_getcapabilities_group_name.py]: Oh, I see. Thank you! This looks much better!

**Comment by tomass** [comment]
@elpaso regarding test case file. It now contains not only GetCapabilities requests, and uniting thing is the new option it is testing and testdata (project files) it is using. So I'd rather leave it as a separate file.
(This also makes me think that one missing thing (or which I could not find) is some meta-file, containing a description of all testdata files - which test cases they are connected to as well as what is they should contain in order for the tests to run - this could be useful to re-use the same test_data for different test cases).

Thank you for your insights and time!

**Pull request merged**
